### DISCUSSION TITLE
Sweep 功能提問

### DISCUSSION THREAD

**Discussion by madduxliu** [JustinSDK/dotSCAD#23] [Q&A]
你好最近剛開始自學opensacd 正在建構電腦風扇的翼型模型
看了很多範例都是指向 SWEEP 的斷面只能用點來建構,但是我的斷面是已經寫好的函式

問題1.   是不是一定要用點來標示斷面才能掃略呢?

問題2.   SWEEP 掃略的路徑是不是也可以用寫好的函式來帶入呢?(因為非直線 會使用二階曲線方程式)

問題3.  現階段嘗試使用LOFT 功能但是好像兩個斷面也是直線混成 是否能修改混成的路徑呢??

**Comment by JustinSDK** [comment]
「斷面是已經寫好的函式」是指什麼？寫好的 `module`？如果你的函式指的是 `function`，OpenSCAD 強制一定會有傳回值，那麼你傳回的是什麼？

sweep 本質上，就是從你提供的幾何資料（路徑、斷面的點等），自動計算出「面」來構成 3D 實體，如果你已經用 `module` 畫出東西了，OpenSCAD 是無法從中取得幾何資訊的。

無論是 sweep 或 loft，兩個斷面間本來就是直線，若要曲線，就是由許多更小的直線接合成看似曲線的東西，應該說，電腦的世界中，本質上沒有真正的曲線這種東西，你所謂的二階曲線方程式，在計算時，也會產生一堆點，點與點間就是直線，只不過夠接近的話，許多點看起來會像是曲線。

**Comment by madduxliu** [comment]
我先在naca6 裡面寫好我要的三個function 再用最上層的function 來執行成我要的斷面 (代替module )

use <naca6.SCAD> //
use <tween_loft.scad>			// Define all functions (this is the main file with full documentation)
include <tween_shapes.scad>	// Define all tween shape geometries
 

 
// ------------------------Parameters 

//-----------------------------------------------

// The lower shape
shape1				= airfoil_data(naca=4412, L = 100, N = 81, open = false);		
shape1Size 		= 1;				// Size of the lower shape
shape1Rotation 	= 0;				// Rotation of the lower shape
shape1Extension 	= 0;				// Extend the profile (space for tube clamp, etc.)
shape1Centroid  	= [0,0];			// Location of center point

// The upper shape
shape2				= airfoil_data(naca=4421, L = 50, N = 81, open = false);		
shape2Size 		= 1;				// Size of the upper shape
shape2Rotation 	= 45;				// RotationSize of the upper shape
shape2Extension 	= 0;				// Extend the profile (space for tube clamp, etc.)
shape2Centroid	= [0,10];			// Location of center point
shape2ExtensionAdjustment	= 0;	// Moves top extension down by n slices.

wallThickness		= 1;				// Wall Thickness - higher values add material but will seal gaps
										// Thickness is added to the exterior diameter of tu be, no effect on solids
				
isHollow 			= 0;				// If 1, create a tube.  If 0, create a solid.

extrusionHeight	= 100;				// Height of the loft

extrusionSlices 	= 30;	
sliceAdjustment	= 0;				// Ensure the slices intersect by this amount, 
										// needed if OpenSCAD is generating more than 2 volumes for an STL file

sliceHeight = extrusionHeight * 1.0 / extrusionSlices;	// Calculate the height of each slice

// Generate the top level part

// Make the container
tweenLoft(shape1, shape1Size, shape1Rotation, shape1Centroid, shape1Extension,
			shape2, shape2Size, shape2Rotation, shape2Centroid, shape2Extension, shape2ExtensionAdjustment,
			extrusionSlices, sliceHeight, sliceAdjustment, wallThickness/2, isHollow);

**Comment by madduxliu** [comment]
以下naca6 的內容


// this is the object
module airfoil(naca=12, L = 100, N = 81, h = 1, open = false)
{
 rotate([180,0,180])
  polygon(points = airfoil_data(naca, L, N, open)); 
}

// this is the main function providing the airfoil data
function airfoil_data(naca=12, L = 100, N = 81, open = false) = 
let(Na = (naca)!=3?NACA(naca):naca)
  let(A = [.2969, -0.126, -.3516, .2843, open?-0.1015:-0.1036])
  [for (b=[-180:360/(N):179.99]) 
    let (x = (1-cos(b))/2)  
    let(yt = sign(b)*Na[2]/.2*(A*[sqrt(x), x, x*x, x*x*x, x*x*x*x])) 
    Na[0]==0?L*[x, yt]:L*camber(x, yt, Na[0], Na[1], sign(b))];  

// helper functions
function NACA(naca) = 
  let (M = floor(naca/1000))
  let (P = floor((naca-M*1000)/100)) 
  [M/100, P/10, floor(naca-M*1000-P*100)/100];  

function camber(x, y, M, P, upper) = 
  let(yc = (x<P)?M/P/P*(2*P*x-x*x): M/(1-P)/(1-P)*(1 - 2*P + 2*P*x -x*x))
  let(dy = (x<P)?2*M/P/P*(P-x):2*M/(1-P)/(1-P)*(P-x))
  let(th = atan(dy))
  [upper ? x - y*sin(th)-1:x + y*sin(th)-1, upper ? yc + y*cos(th):yc - y*cos(th)];

**Comment by madduxliu** [comment]
![image](https://user-images.githubusercontent.com/91867937/135818737-fb3d40bd-ff92-4454-8ff2-e0fb2f69d9f6.png)

**Comment by JustinSDK** [comment]
> 我先在naca6 裡面寫好我要的三個function 再用最上層的function 來執行成我要的斷面 (代替module )

三個 `function` 是指 `airfoil_data`、`NACA`、`camber`？最上層的 `function` 是指 `airfoil`？`airfoil` 是 `module` 定義的啊？你的程式碼中也沒有用到 `airfoil` 這個模組，我不懂你想做什麼！
